RHOAI-4148: Re-adjust the statefulset, route, service based on generateName

harshad16 opened the pull request:
https://issues.redhat.com/browse/RHOAIENG-4148

<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
Re-adjust the statefulset, route, service based on generateName

The setup on Opendatahub allows spin-up of workbenches, via 
- Data Science Project
- Jupyter Tile (application>enabled)

Way of naming in the above methods:
- Data Science Project: users choose to name it 
- Jupyter Tile: `jupyter-nb-<encoded-username>`

Length issue on Jupyter Tile:
- Depending on the username, the dashboard [encode special characters](https://github.com/opendatahub-io/odh-dashboard/blob/0a4fd273bed8af3cd0ccfb1a61a0a64c37d5abde/frontend/src/utilities/notebookControllerUtils.ts#L30) as not special chars are accepted in Kubernetes. 
  - This can cause the length of the username to change with few added chars.
 - Dashboard adds `jupyter-nb-` that is 11 len, [here](https://github.com/opendatahub-io/odh-dashboard/blob/0a4fd273bed8af3cd0ccfb1a61a0a64c37d5abde/backend/src/utils/notebookUtils.ts#L31).
 
Explanation:
- The route URL is created by sub-domain`(<route-name>-<namespace>)` : `<route-name>-<namespace>.<domain>`
- The len of sub-domain needs to be 63 chars. References:
   - https://docs.openshift.com/container-platform/3.11/architecture/networking/routes.html#:~:text=Each%20route%20consists%20of%20a,and%20an%20optional%20security%20configuration.
   - https://access.redhat.com/solutions/3638581
- As jupyter-tiles are created in namespace (rhods-notebooks) this add: 16 chars to sub-domain.
- Following get the limit of len of username: 63-16 = 47 - (len of prefix jupyter-nb i.e 11) = 36.

In ODH username above 36 chars were failing , fixing the Route name to generated using Kubernets generateName method, we can fix the route sub-domain issue. 

Upon further exploring found that labels have length constraint of 63 chars.
- https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
- https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

The controllerRevision, adds a hash that is 8-10Chars, which adds 11 chars to a label of statefulset.
Given a workbench in jupyter-tile, that allows: 
username: 63- (len of prefix jupyter-nb i.e 11) - (len of hash i.e 11) = 41

In ODH username between 36-41 chars were failing in statefulset creation stage , fixing the Statefulset name to generated using Kubernets generateName method, we can fix the controller-revision-hash label.

With this, username can be up to 52 chars. 
as there are other labels, which would fail the 63 chars limit (app, notebook-name labels)

End Result:
Old username limit: max 36
New username limit: max 52

Related-to: 
https://github.com/opendatahub-io/kubeflow/issues/179
https://issues.redhat.com/browse/RHOAIENG-4148

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
TBD

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

harshad16 committed:
Adjust the route url to maintain the sub-domain length less than 63 char

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

harshad16 committed:
Adjust the statefulset naming based on the length

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

harshad16 committed:
Use constant vars for routelengthmax and namespacemax

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

harshad16 committed:
Include the env test case for the route generateName

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

jiridanek commented:
This is good work, but I worry it's not really all that much helpful to fix https://issues.redhat.com/browse/RHOAIENG-4148

For odh/kubeflow, the thing to help with the customer issue, would be to reject Notebook CR creation immediately, instead of getting stuck creating route, or whatever, later.

The largest weight lays on Dashboard which simply cannot put the username into the Notebook CR name, because the usernames can be long.

It is a good change, but it will not resolve RHOAIENG-4148 alone.

Also, it is getting us towards

* https://github.com/kubeflow/kubeflow/pull/7468#issuecomment-2187500924

which is probably a good direction to move towards.

harshad16 committed:
Check limit for entire sub-domain instead of name and namespace

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

jiridanek commented:
Ftr, there's a new feature in rhods operator to let admin change the Jupyterlab namespace, https://issues.redhat.com/browse/RHOAIENG-22096

Default still is

```
 ${DEFAULT_WORKBENCHES_NAMESPACE}=    rhods-notebooks
```

harshad16 commented:
> Ftr, there's a new feature in rhods operator to let admin change the Jupyterlab namespace, https://issues.redhat.com/browse/RHOAIENG-22096
> 
> Default still is
> 
> ```
>  ${DEFAULT_WORKBENCHES_NAMESPACE}=    rhods-notebooks
> ```

Great point, noticing that, I was trying to change the logic for making it more generic :+1:

jiridanek commented:
/lgtm

I've been thinking about this and while I still believe what I said/thought before (the change does not really fix customer issue, change will move our logic further from upstream kubeflow notebooks, the branching logic for the route takes time to understand), I don't see anything wrong with this.

In the interest of making further progress, I think we should merge this. It will fix _some_ scenarios, upstream kubeflow is not really making any changes in the 1.x notebook controller codebase, and the logic is not too bad and worst case chatgpt can help explain if I ever stumble with it.

Ship it!

harshad16 committed:
adjust the finding of route for oauthclient

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

jstourac commented:
Do we want to do something with the failing CI code static analysis check?
```
prefix=components/odh-notebook-controller --timeout=5m] in [/home/runner/work/kubeflow/kubeflow/components/odh-notebook-controller] ...
  Error: components/odh-notebook-controller/controllers/notebook_route.go:106:21: ST1023: should omit type bool from declaration; it will be inferred from the right-hand side (staticcheck)
  	var isGenerateName bool = false
  	                   ^
  Error: components/odh-notebook-controller/controllers/notebook_route.go:236:5: QF1008: could remove embedded field "ObjectMeta" from selector (staticcheck)
  	sa.ObjectMeta.Annotations["serviceaccounts.openshift.io/oauth-redirectreference.second"] = "" +
  	   ^
  2 issues:
  * staticcheck: 2
  ```
AFAIK it's not failing in main?

jstourac commented on components/notebook-controller/controllers/notebook_controller.go:
```diff
@@ -405,17 +430,25 @@ func setPrefixEnvVar(instance *v1beta1.Notebook, container *corev1.Container) {
+	}
+	if isGenerateName {
+		ssObjectMeta = metav1.ObjectMeta{
+			GenerateName: "nb-",
```
How will this look like in the real world? isn't it too generic? I mean - would it make sense to have at least some portion of the actual original `instance.Name` here so it's at least somehow traceable to the original user via `oc get ...` command?

harshad16 replied:
> How will this look like in the real world? isn't it too generic?

As this change would only happen if `isGenerateName` is set, which depends on notebook name to be larger than 52 char. In ODH scenario, it is only possible via jupyter tile , as DS project has slices notebook name above 30 char. 
In jupyter tile, user dont get to set the notebook name on the UI, so on console it would show up as nb-<hash>
user can query them via cli using:
`oc get notebook --all-namespaces -l opendatahub.io/user={username}`
`
or 
`
oc get statefulset --all-namespaces -l statefulset={notebook.name}
`

we could have some part of instance.name, i wasn't sure, what slice should it be prefix or suffix end of instance.name.

harshad16 replied:
Absolutely, i do feel inline with you. 
With this case being exclusive to jupyter tiles, keeping the name sequence could be a way. Perhaps we can do this in next iterations.

daniellutz commented:
I understand and agree with @jiridanek's comments.

Since some of these are mostly due to OCP's own limitations and requirements, this PR will be a great fix to these issues already. Tests are present, only some linter checks appearing on CI, but everything looks good!

/lgtm

harshad16 committed:
Update components/notebook-controller/controllers/notebook_controller.go

Co-authored-by: Jan Stourac <jstourac@redhat.com>

harshad16 commented on components/notebook-controller/controllers/notebook_controller.go:
```diff
@@ -136,16 +139,38 @@ func (r *NotebookReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
+	// controller-service-hash that shouldn't exceed 63 chars.
+	isGenerateName := false
+	if len(instance.Name) > MaxStatefulsetNameLength {
+		log.Info("Notebook name is too long, it should be less than 52 chars")
```
thank you for this, definitely worth a change.

harshad16 committed:
Fix the lint issue
- ST1023: should omit type bool from declaration; it will be inferred from the right-hand side (staticcheck)
- QF1008: could remove embedded field "ObjectMeta" from selector (staticcheck)

Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

jstourac commented:
/lgtm

harshad16 commented:
/retest-required
